lib/ddtrace/contrib/sinatra/tracer.rb

```diff
@@ -32,53 +36,115 @@ def route(verb, action, *)
+            span.set_error(env['sinatra.error']) if response.server_error?
+
+            Sinatra::Env.set_middleware_traced(env, true)
           end
+        end
 
-          app.after do
-            return unless Datadog.configuration[:sinatra][:tracer].enabled
+        # Retrieves of initializes the top-most middleware Sinatra span.
+        # This allows us to only create spans for Sinatra middlewares that
+        # have a matching route.
+        #
+        # If we traced all Sinatra middleware apps, we would have a chain
+        # of nested spans that were not responsible for handling this request,
+        # adding little value to users.
+        def self.fetch_middleware_span!(env, configuration)
+          span = Sinatra::Env.datadog_span(env)
+          return span if span
+
+          tracer = configuration[:tracer]
+          span = tracer.trace(
+            Ext::SPAN_REQUEST,
+            service: configuration[:service_name],
+            span_type: Datadog::Ext::HTTP::TYPE_INBOUND,
+            start_time: Sinatra::Env.middleware_start_time(env)
+          )
+
+          Sinatra::Env.set_datadog_span(env, span)
+          span
+        end
+
+        # Method overrides for Sinatra::Base
+        module Base
+          def render(engine, data, *)
+            tracer = Datadog.configuration[:sinatra][:tracer]
+            return super unless tracer.enabled
+
+            tracer.trace(Ext::SPAN_RENDER_TEMPLATE, span_type: Datadog::Ext::HTTP::TEMPLATE) do |span|
+              span.set_tag(Ext::TAG_TEMPLATE_ENGINE, engine)
+
+              # If data is a string, it is a literal template and we don't
+              # want to record it.
+              span.set_tag(Ext::TAG_TEMPLATE_NAME, data) if data.is_a? Symbol
 
-            span = Sinatra::Env.datadog_span(env)
+              # Measure service stats
+              Contrib::Analytics.set_measured(span)
 
-            unless span
-              Datadog.logger.error('missing request span in :after hook')
-              return
+              super
             end
+          end
 
-            span.resource = "#{request.request_method} #{@datadog_route}"
-            span.set_tag(Ext::TAG_ROUTE_PATH, @datadog_route)
-            span.set_tag(Datadog::Ext::HTTP::STATUS_CODE, response.status)
-            span.set_error(env['sinatra.error']) if response.server_error?
+          # Invoked when a matching route is found.
+          # This method yields directly to user code.
+          def route_eval
+            configuration = Datadog.configuration[:sinatra]
+            tracer = configuration[:tracer]
+
+            return super unless tracer.enabled
+
+            # For initialization of Sinatra middleware span in order
+            # guarantee that the Ext::SPAN_ROUTE span being created below
+            # has the middleware span as its parent.
+            Sinatra::Tracer.fetch_middleware_span!(env, configuration)
```

Comment:
It's interesting to me that we `fetch!` here without actually wanting to use the result at all; if you `fetch`, I'd think you'd actually want to use what you fetched. I think it actually demonstrates an "exploitation" of side effects, rather that using the function for what its meant to do.

This suggests to me that maybe there should be explicit `create` and `get` behaviors? Whether there's still a `fetch` that wraps `create`/`get` is up to you, but ultimately I think I'd like to see this line declare its intent more clearly with something like `create`, so its unambiguous.

Reply:
I added the `!` at the end to show side-effect, and also renamed the method from `get` to `fetch`, but I can rename it further, will do.

Reply:
Extract behaviour into 2 different components.